### ISSUE TITLE
[Enhancement]Loss given by human, optimization given by LLM

### ISSUE THREAD

**Issue by IcyFeather233** [zou-group/textgrad#39]
Hello, I very like this project！

In my use, I often encounter scenarios where both loss calculation and optimization are performed by LLMs. This analytical process heavily relies on the capabilities of the LLM itself. If the LLM's analysis of the loss is not good, then the optimization part will also deviate from our intended goals. 

So is it possible for me to manually provide the loss, but still use the LLM for optimization? If so, what modifications would be needed? I would appreciate a general guidance on the direction of the modifications! Thank you very much!

**Comment by mertyg** [comment]
Yeah we are very excited about this direction and is in our roadmap. Overall, it should not be too difficult to do this. For instance, a hacky way would be using [StringBasedFunction](https://github.com/zou-group/textgrad/blob/main/textgrad/autograd/string_based_ops.py#L43) and instead of actually running something like exact match, taking an input from a user.

However, there should be better ways to do this, and we'll certainly push for this in the near future!

**Comment by beansrowning** [comment]
I'm not confident that I'm doing this how @mertyg suggested, but I had a related need for a custom loss function (non-LLM). 

In my case, I wanted to improve a prompt to repair CSV files that were structurally damaged, so I wrote a loss function where I try to read in the LLM response using Polars and compare against the corrected file. That looked like this:

```py
import re
from io import StringIO
import polars as pl
import textgrad as tg
from textgrad.autograd import Module, StringBasedFunction

class CSVLoss(Module):
    """
    A Hacky CSV comparison loss FN
    """

    def __init__(self) -> None:
        super().__init__()
        self.call_fn = StringBasedFunction(
            fn=self._eval_df, function_purpose="Evaluate CSV response to ground truth"
        )

    @staticmethod
    def extract_csv(input_str: str) -> str:
        """
        Extract just the CSV content from a body of text, assuming it's properly denoted within a ```csv block
        """
        csv_re = re.compile(r"```[cC]sv\n(.+?)```", re.DOTALL)
        return csv_re.findall(input_str)[0]

    @classmethod
    def _eval_df(cls, input_val: tg.Variable, truth: tg.Variable) -> str:
        """
        Evaluate if model response x matches the expected label y
        """
        try:
            x_df = pl.read_csv(StringIO(cls.extract_csv(input_val.get_value())))
        except pl.exceptions.ComputeError as e:
            if "found more fields than defined" in str(e):
                out = f"Parsing input CSV via `polars.read_csv(<input_csv>)` failed because there were >=1 rows with more fields than columns in the CSV header"
                return out
        except Exception as e:
            # Early return if this still fails to parse
            out = f"Parsing input CSV via `polars.read_csv(<input_csv>)` failed with following exception: {str(e)}"
            return out

        y_df = pl.read_csv(StringIO(cls.extract_csv(truth.get_value())))

        if x_df.equals(y_df):
            out = "Answer Correct."
        else:
            out = "Answer parses correctly via `polars.read_csv(<input_csv>)`, but doesn't match expected value"

        return out

    def forward(self, **kwargs):
        return self.call_fn(inputs=kwargs)
```

And that drops right in to a standard training loop:

```py
def main():
    loss_fn = CSVLoss()
    # ... <snip>
    # Run for N epochs (+1 for starting place)
    for i in range(N_EPOCHS + 1):
        print(f"=== Round {i} ===============")
        # Train
        optimizer.zero_grad()
        answer = model(question)
        loss = loss_fn(input_val=answer, truth=correct)
        loss.backward()
        print(f"loss: {loss}")
        optimizer.step()
```

Hopefully, that helps someone browsing here. I'm happy to help contribute too if there's need/desire.

**Comment by mertyg** [comment]
This is a very interesting and cool use case ! I really like this example. Do you want to contribute a PR with this loss function and an example notebook? I'd love to have this!

Thank you so much @beansrowning !

**Comment by chrico-bu-uab** [comment]
I am researching whether LLMs can correctly identify cancer recurrence severity in clinical text. I tried to mimic the code from @beansrowning (thanks!) but am not having much luck. Unfortunately, depending on which model(s) I use, the system prompt:
- barely changes (with no or "negative" impact on the loss), or
- changes too much (to the point where it's completely divorced from the actual task per `starting_prompt`), or
- includes "hints" based on a particular patient it had encountered in my data, hints that only apply to that patient.

I was hoping for the process to add some sort of nuance to the definitions that I spell out in `TERMS`, but this never is the case - it either throws out the terms or leaves them untouched.

Is there a way I could modify the role descriptions and/or other instructions such that more thoughtful and desirable changes to the system prompt are elicited? Or maybe I wrote the loss function incorrectly?

Note: I've only tried this on smaller models, mostly llama3.2 3b with GPT-4o-mini as the backend.

Also, it would be nice if there were a way to quantify the loss and track its change over epochs.

Below is my code.
```python
TERMS = ("\n"
         "\n"
         "### Degrees of Cancer Recurrence Severity:\n"
         "\n"
         "1. **Not Recurrence**:\n"
         "   - Cancer is either still present after treatment (**persistent disease**) or remains absent with no "
         "evidence of disease (**remission**). For our purposes, both scenarios are considered \"not recurrence.\"\n"
         "\n"
         "2. **Local Recurrence**:\n"
         "   - Cancer returns at the **same location** or very close to where it first developed.\n"
         "\n"
         "3. **Regional Recurrence**:\n"
         "   - Cancer returns in **nearby lymph nodes** or tissues adjacent to the original tumor site.\n"
         "\n"
         "4. **Distant Recurrence**:\n"
         "   - Also known as **metastatic recurrence**; cancer returns in a **different part of the body** far from "
         "the original site.")

TEXT_CAVEAT = ("\n\n**Note**: The provided text may not explicitly mention the correct recurrence severity, and might "
               "even state a wrong one. You'll need to infer the correct severity based on the information provided.")


class SeverityLoss(Module):
    def __init__(self) -> None:
        super().__init__()
        self.call_fn = StringBasedFunction(
            fn=self._eval_df,
            function_purpose=(
                    "Evaluate the model's prediction for cancer recurrence severity against the ground truth."
                    " Assign a penalty based on the distance between the predicted severity and the actual severity "
                    "on the recurrence scale."
                    + TERMS
            ),
        )

    @classmethod
    def _eval_df(cls, input_val: tg.Variable, truth: tg.Variable) -> str:
        x = input_val.value.lower()
        y = truth.value.lower()

        match = re.search('|'.join(SEVERITIES), x)
        if match is None:
            return "The model did not predict any of the possible cancer recurrence severities."

        answer = x[match.start():match.end()]
        pred_index = severity_indices[answer]
        true_index = severity_indices[y]

        distance = abs(pred_index - true_index)
        if distance == 0:
            return f"The model correctly predicted the cancer recurrence severity as {y}."
        else:
            return (
                f"The model's prediction '{answer}' is {distance} level(s) away from the actual severity '{y}'. "
                f"Assign a penalty proportional to this distance."
            )

    def forward(self, **kwargs):
        return self.call_fn(inputs=kwargs)


def improve_system_prompt(starting_prompt, question_strs, answer_strs, eval_fn):
    system_prompt = tg.Variable(
        starting_prompt,
        requires_grad=True,
        role_description="system"
    )
    model = tg.BlackboxLLM(weak_engine, system_prompt=system_prompt)
    optimizer = tg.TGD(parameters=list(model.parameters()), verbose=3)
    questions = [tg.Variable(question_str, role_description="question", requires_grad=False)
                 for question_str in question_strs]
    answers = [tg.Variable(answer_str, role_description="answer", requires_grad=False)
               for answer_str in answer_strs]
    for _ in range(N_EPOCHS):
        optimizer.zero_grad()
        losses = [eval_fn(input_val=model(question), truth=answer)
                  for question, answer in tqdm(list(zip(questions, answers)), desc="Calculating losses")]
        total_loss = tg.sum(losses)
        total_loss.backward()
        optimizer.step()
    return system_prompt.value


...
        starting_prompt = (
                              "As a medical expert, determine the most accurate cancer recurrence severity for the "
                              "patient based on the provided information. Your assessment should strictly be one of "
                              f"the following: {add_or(', '.join(SEVERITIES))}. Provide a concise justification for "
                              "your choice, referencing specific details from the patient's information without "
                              "introducing new data."
                          ) + TERMS + TEXT_CAVEAT
        system_prompt = improve_system_prompt(
            starting_prompt,
            support["text"].values,
            support["Severity"].values,
            SeverityLoss()
        )
...
```
